agent/structs/acl.go

```diff
@@ -1,155 +1,773 @@
 package structs
 
 import (
+	"encoding/binary"
 	"errors"
+	"fmt"
+	"hash/fnv"
+	"sort"
+	"strings"
 	"time"
 
 	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/sentinel"
+	"golang.org/x/crypto/blake2b"
+)
+
+type ACLMode string
+
+const (
+	// ACLs are disabled by configuration
+	ACLModeDisabled ACLMode = "0"
+	// ACLs are enabled
+	ACLModeEnabled ACLMode = "1"
+	// DEPRECATED (ACL-Legacy-Compat) - only needed while legacy ACLs are supported
+	// ACLs are enabled and using legacy ACLs
+	ACLModeLegacy ACLMode = "2"
+	// DEPRECATED (ACL-Legacy-Compat) - only needed while legacy ACLs are supported
+	// ACLs are assumed enabled but not being advertised
+	ACLModeUnknown ACLMode = "3"
 )
 
 // ACLOp is used in RPCs to encode ACL operations.
 type ACLOp string
 
+type ACLTokenIDType string
+
+const (
+	ACLTokenSecret   ACLTokenIDType = "secret"
+	ACLTokenAccessor ACLTokenIDType = "accessor"
+)
+
+type ACLPolicyIDType string
+
 const (
-	// ACLBootstrapInit is used to perform a scan for existing tokens which
-	// will decide whether bootstrapping is allowed for a cluster. This is
-	// initiated by the leader when it steps up, if necessary.
-	ACLBootstrapInit = "bootstrap-init"
+	ACLPolicyName ACLPolicyIDType = "name"
+	ACLPolicyID   ACLPolicyIDType = "id"
+)
+
+const (
+	// All policy ids with the first 120 bits set to all zeroes are
+	// reserved for builtin policies. Policy creation will ensure we
+	// dont accidentally create them when autogenerating uuids.
+
+	// This policy gives unlimited access to everything. Users
+	// may rename if desired but cannot delete or modify the rules
+	ACLPolicyGlobalManagementID = "00000000-0000-0000-0000-000000000001"
+	ACLPolicyGlobalManagement   = `
+acl = "write"
+agent_prefix "" {
+	policy = "write"
+}
+event_prefix "" {
+	policy = "write"
+}
+key_prefix "" {
+	policy = "write"
+}
+keyring = "write"
+node_prefix "" {
+	policy = "write"
+}
+operator = "write"
+query_prefix "" {
+	policy = "write"
+}
+service_prefix "" {
+	policy = "write"
+	intention = "write"
+}
+session_prefix "" {
+	policy = "write"
+}`
+
+	// This is the policy ID for anonymous access. This is configurable by the
+	ACLTokenAnonymousID = "00000000-0000-0000-0000-000000000002"
+)
 
-	// ACLBootstrapNow is used to perform a one-time ACL bootstrap operation on
-	// a cluster to get the first management token.
-	ACLBootstrapNow = "bootstrap-now"
+func ACLIDReserved(id string) bool {
+	return strings.HasPrefix(id, "00000000-0000-0000-0000-0000000000")
```

Comment:
Should we also reserve `ACLTokenAnonymousID` and `ACLPolicyGlobalManagementID` here too?

Reply:
It is, the prefix being checked here is for 120 bits of 0's. So we are reserving 256 IDs for our usage.

Reply:
Oh cool. that was super non-obvious to me even though I noted it was Prefix match just because I didn't count the zeros and it looks like a full UUID. If you didn't already maybe comment that?